src/parcsr_ls/par_mgr_rap.c

```diff
@@ -0,0 +1,602 @@
+                                                    ordering, method, max_elmts, A_H_ptr);
+   }
+   else
+#endif
+   {
+      hypre_MGRBuildNonGalerkinCoarseOperatorHost(A_FF, A_FC, A_CF, A_CC, Wp, Wr,
+                                                  fine_blk_dim, coarse_blk_dim,
+                                                  ordering, method, max_elmts, A_H_ptr);
+   }
+
+   return hypre_error_flag;
+}
+
+/*--------------------------------------------------------------------------
+ * hypre_MGRBuildCoarseOperator
+ *--------------------------------------------------------------------------*/
+
+HYPRE_Int
+hypre_MGRBuildCoarseOperator(void                *mgr_vdata,
+                             hypre_ParCSRMatrix  *A_FF,
+                             hypre_ParCSRMatrix  *A_FC,
+                             hypre_ParCSRMatrix  *A_CF,
+                             hypre_ParCSRMatrix **A_CC_ptr,
+                             hypre_ParCSRMatrix  *Wp,
+                             hypre_ParCSRMatrix  *Wr,
+                             HYPRE_Int            level)
+{
+   hypre_ParMGRData      *mgr_data = (hypre_ParMGRData*) mgr_vdata;
+   hypre_ParCSRMatrix    *A  = (mgr_data -> A_array)[level];
+   hypre_ParCSRMatrix    *P  = (mgr_data -> P_array)[level];
+   hypre_ParCSRMatrix    *R  = (mgr_data -> R_array)[level];
+   hypre_ParCSRMatrix    *RT = (mgr_data -> RT_array)[level];
+   hypre_ParCSRMatrix    *A_CC = *A_CC_ptr;
+
+   HYPRE_Int             *blk_dims = (mgr_data -> block_num_coarse_indexes);
+   HYPRE_Int              block_size = (mgr_data -> block_size);
+   HYPRE_Int              method = (mgr_data -> mgr_coarse_grid_method)[level];
+   HYPRE_Int              num_coarse_levels = (mgr_data -> max_num_coarse_levels);
+   HYPRE_Int              ordering = (mgr_data -> set_c_points_method);
+   HYPRE_Int              max_elmts = (mgr_data -> nonglk_max_elmts)[level];
+   HYPRE_Real             threshold = (mgr_data -> truncate_coarse_grid_threshold);
+
+   hypre_ParCSRMatrix    *AP, *RAP, *RAP_c;
+   HYPRE_Int              fine_blk_dim = (level) ? blk_dims[level - 1] - blk_dims[level] :
+                                           block_size - blk_dims[level];
+   HYPRE_Int              coarse_blk_dim = blk_dims[level];
+
+   HYPRE_ANNOTATE_FUNC_BEGIN;
+   hypre_GpuProfilingPushRange("RAP");
+
+   if (!method)
+   {
+      /* Galerkin path */
+      if (Wr && !Wp)
+      {
+         /* Prolongation is the injection operator (Wp == NULL) and
+            Restriction is not the injection operator (Wr != NULL) */
+         RAP_c = hypre_ParCSRMatMat(Wr, A_FC);
+         hypre_ParCSRMatrixAdd(1.0, A_CC, -1.0, RAP_c, &RAP);
+      }
+      else if (RT)
+      {
+         RAP = hypre_ParCSRMatrixRAPKT(RT, A, P, 1);
+      }
+      else if (R)
+      {
+         AP  = hypre_ParCSRMatMat(A, P);
+         RAP = hypre_ParCSRMatMat(R, AP);
+         hypre_CSRMatrixReorder(hypre_ParCSRMatrixDiag(RAP));
```

Comment:
Remind me why we are reordering here?

Reply:
This is from some time ago and I'm not 100% sure, but I believe it's because `hypre_ParCSRMatMat` doesn't guarantee that the diagonal entry goes first. So the reorder is to guarantee that. @liruipeng is this correct?